fix: graph algorithm improvements

ayaankhan98 opened the pull request:
#### Description of Change
<!--
Thank you for your Pull Request. Please provide a description above and review
the requirements below.

Contributors guide: https://github.com/TheAlgorithms/C-Plus-Plus/CONTRIBUTING.md
-->
issue related to this #909
- Code quality Improvement over graph Algorithms
- Added Documentation
- Added CMakeLists.txt

#### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#New-File-Name-guidelines)
- [x] Added tests and example, test must pass
- [x] Added documentation so that the program is self-explanatory and educational - [Doxygen guidelines](https://www.doxygen.nl/manual/docblocks.html)
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes: <!-- Please add a one-line description for developers or pull request viewers -->

ayaankhan98 commented:
Do not merge this, i will be improving all files one by one. it will be easy to review single file at a time

kvedala commented:
> Do not merge this, i will be improving all files one by one. it will be easy to review single file at a time

You can mark the pull request as "draft" and work on it. Once it is ready, you can mark it as ready for review.

kvedala commented:
### Note
If There are files to be *renamed*, use the `git move` operation instead of deleting and adding a file

ayaankhan98 commented:
> ### Note
> If There are files to be _renamed_, use the `git move` operation instead of deleting and adding a file. This will ensure that the file history is maintained. Otherwise, it may lead to a discontinuity in the file history.

i didn't know about git move. thanks for telling, i will apply this if needed again.

kvedala commented:
> > ### Note
> > If There are files to be _renamed_, use the `git move` operation instead of deleting and adding a file. This will ensure that the file history is maintained. Otherwise, it may lead to a discontinuity in the file history.
> 
> i didn't know about git move. thanks for telling, i will apply this if needed again.

This is with regards to `bfs.cpp` file as well. If deleted, the contributor's history will be lost. 
Thanks for understanding.

ayaankhan98 commented:
> > > ### Note
> > > If There are files to be _renamed_, use the `git move` operation instead of deleting and adding a file. This will ensure that the file history is maintained. Otherwise, it may lead to a discontinuity in the file history.
> > 
> > 
> > i didn't know about git move. thanks for telling, i will apply this if needed again.
> 
> This is with regards to `bfs.cpp` file as well. If deleted, the contributor's history will be lost.
> Thanks for understanding.

yes i got that

kvedala commented:
for BFS, the object oriented implementation of `bfs.cpp` is more lucrative for adoption than functional implementation of the new file `breadth_first_search.cpp`. We should keep them both unless the first one is erroneous.

ayaankhan98 commented:
> for BFS, the object oriented implementation of `bfs.cpp` is more lucrative for adoption than functional implementation of the new file `breadth_first_search.cpp`. We should keep them both unless the first one is erroneous.

you are right, the `bfs.cpp` implementaion was not bad. the reason i choosed the functional approch is that as in several files under graph we have object oriented approch then we will define class graph form each file and in the documentaion we will have repeated methods of graph class, one thing that we can do to fix this is we can define `graph.hpp` and include this in other files but this will separate the the matter in two files and for newbie it will be hard to understand the algorithm properly. so i think functonal approach is good for this purpose, it will make the implementaion easily understandable and we will have consistency in documentation also.
Other drawback of the approch using `graph.hpp` that i mentioned earlier is that if we define class graph and it's methods seprately then there is no benefit of this because the algorithms such as bfs and dfs are not always used as such we always slightly tweak them to achieve the desired behaviour such as if we wish to count the number of connected compoents then we can use dfs to do this but in this case we will not use the simple dfs we will tweak it a bit some where in between to keep count of the connected components, similiary in case if we need to count the pre order or post order of vertices then we can use dfs to get this but here also we will tweak the dfs slightly to achieve our purpose so a general implementation of dfs defiined in `graph.cpp` corrosponding to `graph.hpp` will not be beneficial in these case.

kvedala commented:
> > for BFS, the object oriented implementation of `bfs.cpp` is more lucrative for adoption than functional implementation of the new file `breadth_first_search.cpp`. We should keep them both unless the first one is erroneous.
> 
> you are right, the `bfs.cpp` implementaion was not bad. the reason i choosed the functional approch is that as in several files under graph we have object oriented approch then we will define class graph form each file and in the documentaion we will have repeated methods of graph class, one thing that we can do to fix this is we can define `graph.hpp` and include this in other files but this will separate the the matter in two files and for newbie it will be hard to understand the algorithm properly. so i think functonal approach is good for this purpose, it will make the implementaion easily understandable and we will have consistency in documentation also.
> Other drawback of the approch using `graph.hpp` that i mentioned earlier is that if we define class graph and it's methods seprately then there is no benefit of this because the algorithms such as bfs and dfs are not always used as such we always slightly tweak them to achieve the desired behaviour such as if we wish to count the number of connected compoents then we can use dfs to do this but in this case we will not use the simple dfs we will tweak it a bit some where in between to keep count of the connected components, similiary in case if we need to count the pre order or post order of vertices then we can use dfs to get this but here also we will tweak the dfs slightly to achieve our purpose so a general implementation of dfs defiined in `graph.cpp` corrosponding to `graph.hpp` will not be beneficial in these case.

i think the premise for this is:
> the thought to create a .hpp file
However, that is not necessary. There are a lot of good implementations without an hpp file corresponding to them. The classes like in the existing `bfs.cpp` file are fine opjects implemented in the same cpp file. The repository is providing implementation examples only and how users want to incorporate them into their project is to their discretion and independent of this repo. Hope that clarified your concerns 😄

ayaankhan98 commented:
is this type of document is consitent? we will have repeated stuff like this if we implement class graph in all files
https://thealgorithms.github.io/C-Plus-Plus/dd/d9b/classgraph.html

kvedala commented:
> is this type of documentation is consitent? we will have repeated stuff like this if we implement class graph in all files
> https://thealgorithms.github.io/C-Plus-Plus/dd/d9b/classgraph.html
> 
> will this not create confusion to the readers?

You are absolutely correct. Hence, I started grouping the functions, using namespaces and more descriptive variable names to address the problem.

ayaankhan98 commented:
> > is this type of documentation is consitent? we will have repeated stuff like this if we implement class graph in all files
> > https://thealgorithms.github.io/C-Plus-Plus/dd/d9b/classgraph.html
> > will this not create confusion to the readers?
> 
> You are absolutely correct. Hence, I started grouping the functions, using namespaces and more descriptive variable names to address the problem.

great, so shall i continue with the functional approch and group them into a namespace?

kvedala commented:
> > > is this type of documentation is consitent? we will have repeated stuff like this if we implement class graph in all files
> > > https://thealgorithms.github.io/C-Plus-Plus/dd/d9b/classgraph.html
> > > will this not create confusion to the readers?
> > 
> > 
> > You are absolutely correct. Hence, I started grouping the functions, using namespaces and more descriptive variable names to address the problem.
> 
> great, so shall i continue with the functional approch and group them into a namespace?

yes, but *undelete* the original implementation and preferably, update it to the repo standards. Your contribution is an additional algorithm

ayaankhan98 commented:
> > > > is this type of documentation is consitent? we will have repeated stuff like this if we implement class graph in all files
> > > > https://thealgorithms.github.io/C-Plus-Plus/dd/d9b/classgraph.html
> > > > will this not create confusion to the readers?
> > > 
> > > 
> > > You are absolutely correct. Hence, I started grouping the functions, using namespaces and more descriptive variable names to address the problem.
> > 
> > 
> > great, so shall i continue with the functional approch and group them into a namespace?
> 
> yes, but _undelete_ the original implementation and preferably, update it to the repo standards. Your contribution is an additional algorithm

okay no problem simply we are going to maintain both types of implementaions object oriented as well as functional

kvedala commented:
> okay no problem simply we are going to maintain both types of implementaions object oriented as well as functional

Yes, as a repo for algorithm implementations, there could be more than one implementation of one or more algorithms for a problem.
I provided `shell_sort1.cpp` as one of my first contributions to this repo :)

ayaankhan98 commented:
> Also, keep in mind that functional implementations might be more appropriate and efficient with C than with C++.

yes this was also one of the reason to choose functional pattern for graphs

kvedala commented:
> > Also, keep in mind that functional implementations might be more appropriate and efficient with C than with C++.
> 
> yes this was also one of the reason to choose functional pattern for graphs

See for example the implementations for Adaline Machine Learning Algorithm
* C++ OOP implementation [link](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/machine_learning/adaline_learning.cpp)
* C Functional implementation [link](https://github.com/TheAlgorithms/C/blob/master/machine_learning/adaline_learning.c)

kvedala commented:
> * C++ OOP implementation [link](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/machine_learning/adaline_learning.cpp)
> * C Functional implementation [link](https://github.com/TheAlgorithms/C/blob/master/machine_learning/adaline_learning.c)

See the [corresponding PR in C repo](https://github.com/TheAlgorithms/C/pull/555)

kvedala commented on graph/breadth_first_search.cpp:
```diff
@@ -77,8 +77,8 @@ void addEdge(std::vector<std::vector<int>> &adj, size_t u, size_t v) {
-void beadth_first_search(std::vector<std::vector<int>> &adj, size_t start) {
-    size_t vertices = adj.size();
+void beadth_first_search(std::vector<std::vector<int>> *adj, size_t start) {
+    size_t vertices = (*adj).size();
```
for pointers, we use 
```suggestion
    size_t vertices = adj->size();
```

kvedala commented:
Well formatted and well-written code 👍 
Is it ready for review?

ayaankhan98 commented:
> Well formatted and well-written code 
> Is it ready for review?

if you wish we can merge these files and i will imporve rest of files in another PR.

kvedala commented on graph/CMakeLists.txt:
```diff
@@ -0,0 +1,18 @@
+    endif()
+    install(TARGETS ${testname} DESTINATION "bin/graph")
+
+endforeach( testsourcefile ${APP_SOURCES} )
```
```suggestion
endforeach( testsourcefile ${APP_SOURCES} )

```
missing `newline` at EOF

kvedala commented on graph/breadth_first_search.cpp:
```diff
@@ -0,0 +1,135 @@
+ * @param start vertex from where traversing starts
+ *
+ */
+void beadth_first_search(std::vector<std::vector<int>> *adj, size_t start) {
```
performance wise, if `adj` is not being modified, the better function argument would be:
```suggestion
void beadth_first_search(const std::vector<std::vector<int>> &adj, size_t start) {
```

kvedala commented on graph/dfs.cpp:
```diff
@@ -1,26 +1,132 @@
+ * @param visited already visited vertices
+ *
+ */
+void explore(std::vector<std::vector<size_t>> *adj, size_t v,
```
since `adj` is not being modified, the optimal code would be to have
```suggestion
void explore(const std::vector<std::vector<size_t>> &adj, size_t v,
```

kvedala commented on graph/dfs.cpp:
```diff
@@ -1,26 +1,132 @@
+    size_t vertices = adj->size();
+
+    std::vector<bool> visited(vertices, false);
+    explore(&(*adj), start, &visited);
```
redundant
```suggestion
    explore(adj, start, &visited);
```

kvedala commented on graph/dfs.cpp:
```diff
@@ -1,26 +1,132 @@
+ * @param start vertex from where DFS starts traversing.
+ *
+ */
+void depth_first_search(std::vector<std::vector<size_t>> *adj, size_t start) {
```
since `adj` is not being modified,
```suggestion
void depth_first_search(const std::vector<std::vector<size_t>> &adj, size_t start) {
```

kvedala commented on graph/breadth_first_search.cpp:
```diff
@@ -0,0 +1,135 @@
+    while (!tracker.empty()) {
+        size_t vertex = tracker.front();
+        tracker.pop();
+        std::cout << vertex + 1 << " ";
```
It would be better if the function is made modular by returning the result instead of printing it.
The return value can then be utilized by a test function and make the code more manageable

ayaankhan98 replied:
yes you are right but i had not done this beacuse the purpose of these files is only to demonstrate the traversal techinque used to traverse the graphs.

another reason to not to return the value is if i return the value `v+1` from here then the control of program is transfered to the calling function that is the main funcition in this case and one the control is transfered to main function without completely executing the algorithm we cannot get back to the ```breadth_first_search()``` function.

for test ulility what we can do is that we can push these values into a vector or in some other similar structure and then we can test this vector in our tests, but for this too either we need to pass an extra vector by reference as the function argument or we have to maintain a vector globally.

that's why i dropped the idea of writing tests in these traversal algorithms.

kvedala replied:
got it. how about returning an array instead? I am pretty sure traversing a tree can be beneficial in robotics, for example

ayaankhan98 replied:
yes not only in robotics, infact there are a lot of applications of these traversal algorithms.

let's suppose if we return an array there could be two possible approches
- we can pass an array in the breadth_first_search() funciton to fill it with required data and then return this
- and other one is we can maintain a global array then there is no need to returning any array

i will favour the second point because if we choose this then this will not modify the function declaration that is `void breadth_first_seach(parameters...)` and this seems reasonable as the algorithm itself does not return any value it just traverses the graph and keeps the testing part separate but in the first approch we need to modify the function header to return an `int*` pointing to array and it might be a point of confusion for readers.

ayaankhan98 replied:
@kvedala what's your opinion?

kvedala replied:
instantiate a local variable of type `std::vector<T>`.
In this loop, the resulting elements can be appended to it and at the end can be returned. The `std::vector` and `std::array` classes have copy contracts and hence can be safely assigned to in the calling function.

ayaankhan98 replied:
> instantiate a local variable of type `std::vector<T>`.
> In this loop, the resulting elements can be appended to it and at the end can be returned. The `std::vector` and `std::array` classes have copy contracts and hence can be safely assigned to in the calling function.
> Thus, any global variable can be avoided

okay i was thinking that this local variable will get destroyed as soon as the breadth_first_search function smashes from the calling stack.
i was not aware of this that this is the case with c-style arrays.

kvedala replied:
Just to confirm, what I am suggesting is something like so:
```cpp
std::vector<T> beadth_first_search(std::vector<std::vector<int>> *adj, size_t start) {
```
if the resultant size is unknown. If the resulting size is known, then
```cpp
std::array<T, N> beadth_first_search(std::vector<std::vector<int>> *adj, size_t start) {
```
will be more efficient

ayaankhan98 replied:
resultant size is unknown because we are taking input from the used so i used ```std::vector```
@kvedala have a look at the implementaion

ayaankhan98 commented:
> looks great. I have not executed and tested the codes, however.
> Please take a look at the suggestions and let know if that is doable.
> Thank you.
> Good job

for the matter for execution all the code that has been commited so far is executing fine on ```ubuntu 20.04``` with ```g++ 9.3.0``` and ```clang++ 10.0.0```

rest all the suggestions are valuable i will fix them and commit them as soon as possible.

kvedala commented on graph/breadth_first_search.cpp:
```diff
@@ -0,0 +1,197 @@
+    }
+    return result;
+}
+}  // namespace graph
```
```suggestion
```
extra closing brace to be removed

kvedala replied:
Ah, you need to add 
`add_subdirectory(graph)` to `CMakeLists.txt` file in the root folder.

ayaankhan98 replied:
> extra closing brace to be removed. how did it even compile?

oh after sucessfully checking i pull the changes then a merge conflict occured, i think this extra brace is left due to that merge conflict, after fixing merge conflict i did not recomplied the code.
well thanks for pointing this out.

kvedala commented:
hmm. A lot of older files are not compiling. This is why I think I skipped this folder 😟

ayaankhan98 commented:
> hmm. A lot of older files are not compiling. This is why I think I skipped this folder

why is it so?

kvedala commented:
> > hmm. A lot of older files are not compiling. This is why I think I skipped this folder
> 
> why is it so?

Your code was correct until 7871ed8
04c92cf actually breaks *your* code.

kvedala commented:
Step 1 - revert back to your working code 7871ed8

ayaankhan98 commented:
> Step 1 - revert back to your compiling code [7871ed8](https://github.com/TheAlgorithms/C-Plus-Plus/commit/7871ed852c527983e7a2ed3b0d0d02b21dbe851e)

yeah i got it there are lot of files in graph dir which are very poorly written those files even does not including proper header
i tried builing code on my local system. files are written very poorly

ayaankhan98 commented:
should we fix this in this PR or should i remove dir graph from root CMakeLists.txt and merge this? and fix one file at a time in separate PR and at last we can modify root CMakeLists.txt.

kvedala commented:
> should we fix this in this PR or should i remove dir graph from root CMakeLists.txt and merge this? and fix one file at a time in separate PR and at last we can modify root CMakeLists.txt.

good idea. Let us see if the following sequence of steps make sense
1. remove Cmake changes from this PR - keep only the two cpp file changes. Then we keep this on hold.
2. create a new branch *from master*
3. commit the cmake changes here 
4. Create a draft PR from this new branch to master
5. fix the compilation issues one file at a time to this new PR
6. Once all is well and merged, then rebase this PR to the new master.

ayaankhan98 commented:
> > should we fix this in this PR or should i remove dir graph from root CMakeLists.txt and merge this? and fix one file at a time in separate PR and at last we can modify root CMakeLists.txt.
> 
> good idea. Let us see if the following sequence of steps make sense
> 
> 1. remove Cmake changes from this PR - keep only the two cpp file changes. Then we keep this on hold.
> 2. create a new branch _from master_
> 3. commit the cmake changes here
> 4. Create a draft PR from this new branch to master
> 5. fix the compilation issues one file at a time to this new PR
> 6. Once all is well and merged, then rebase this PR to the new master.

it sounds good.
but can we keep the CMake file of graph dir with this PR because it is not breaking the code?

kvedala commented:
> > > should we fix this in this PR or should i remove dir graph from root CMakeLists.txt and merge this? and fix one file at a time in separate PR and at last we can modify root CMakeLists.txt.
> > 
> > 
> > good idea. Let us see if the following sequence of steps make sense
> > 
> > 1. remove Cmake changes from this PR - keep only the two cpp file changes. Then we keep this on hold.
> > 2. create a new branch _from master_
> > 3. commit the cmake changes here
> > 4. Create a draft PR from this new branch to master
> > 5. fix the compilation issues one file at a time to this new PR
> > 6. Once all is well and merged, then rebase this PR to the new master.
> 
> it sounds good.
> but can we keep the CMake file of graph dir with this PR because it is not breaking the code?

We will be adding it to the new PR. Keeping it here as well will break the code. To undo, simply commit deleted file using `git rm --cached graph/CMakeLists.txt` command and commit it.

ayaankhan98 commented:
@kvedala i think it is fine now. isn't it?

kvedala commented on CMakeLists.txt:
```diff
@@ -76,7 +76,6 @@ add_subdirectory(probability)
 add_subdirectory(data_structures)
 add_subdirectory(machine_learning)
 add_subdirectory(numerical_methods)
```
```suggestion
add_subdirectory(numerical_methods)

```
There was a one line space between the two. I have that for readability. This will ensure there are no merge conflicts.

kvedala commented on graph/CMakeLists.txt:
```diff
@@ -0,0 +1,19 @@
+# If necessary, use the RELATIVE flag, otherwise each source file may be listed
```
this file should be deleted as discussed and moved to a new branch.

kvedala commented:
Looks good now. As per discussion, shall we  hold off the merge until the next pull request?

ayaankhan98 commented:
> Looks good now. As per discussion, shall we hold off the merge until the next pull request?

yeah sure.

ayaankhan98 commented:
@kvedala  can we close this PR?, i am merging files of this PR into #954

kvedala commented:
> @kvedala can we close this PR?, i am merging files of this PR into #954

Sounds good 👍 
But, what do you think of the [discussion above](https://github.com/TheAlgorithms/C-Plus-Plus/pull/932#issuecomment-655143941)?

ayaankhan98 commented:
> > @kvedala can we close this PR?, i am merging files of this PR into #954
> 
> Sounds good 
> But, what do you think of the [discussion above](https://github.com/TheAlgorithms/C-Plus-Plus/pull/932#issuecomment-655143941)?

i had added the CMake compilation in that PR and also fixed the breaking code. but some of the files are not documented well, i think it will be better if we merge that PR after i push the code that will not break the repo and let other people give chance to fix the documentation as i am unable to work for some reason for sometime.

kvedala commented:
> > > @kvedala can we close this PR?, i am merging files of this PR into #954
> > 
> > 
> > Sounds good
> > But, what do you think of the [discussion above](https://github.com/TheAlgorithms/C-Plus-Plus/pull/932#issuecomment-655143941)?
> 
> i had added the CMake compilation in that PR and also fixed the breaking code, so closing this will be a good option.

oh ok.
